src/components/MapController/DualMapController/DualMapController.js

```diff
@@ -123,123 +131,77 @@ export default class MapController extends React.Component {
-    // in state, but doesn't select (or store in state) a specific file with a
+    // maps for specific instances.
+    // an 'instance' is a variable + emissions + model + period + run combination. 
+    // Timestamps for an instance may be spread across up to three files 
+    // (one annual, one seasonal, one monthly). DualMapController implicitly receives
+    // the variable, emissions scenario, and model parameters via its metadata props,
+    // which have been filtered on those parameters. It selects a period and run and 
+    // stores them in state, but doesn't select (or store in state) a specific file with a
     // specific unique_id until rendering, when it needs to pass an exact file
-    // and timestamp to the viewer component CanadaMap.
-    // The variable and the comparand may have  different available timestamps.
-
-    const { start_date, end_date, ensemble_member } = dataset;
-
-    // generate the list of available times for one or two variable+run+period combinations.
-    // which may include multiple files of different time resolutions
-    // (annual, seasonal, or monthly).
-    let datasets = _.filter(props.meta,
-      { ensemble_member, start_date, end_date });
-
-    // if there is a comparison variable, get times from its datasets too
-    if (this.hasValidData('comparand', props)) {
-      datasets = datasets.concat(_.filter(props.comparandMeta,
-        { ensemble_member, start_date, end_date }));
-    }
-
-    const timesPromises =
-      datasets.map(ds => getTimeMetadata(ds.unique_id));
-
-    Promise.all(timesPromises).then(responses => {
-      let variableTimes = {};
-      let comparandTimes = {};
-
-      for (let i = 0; i < responses.length; i++) {
-        let id = Object.keys(responses[i].data)[0];
-        for (let timeidx in responses[i].data[id].times) {
-          var idxString = JSON.stringify({
-            timescale: responses[i].data[id].timescale,
-            timeidx,
-          });
-
-          // This assumes only one variable per file.
-          const variable = Object.keys(responses[i].data[id].variables)[0];
-          if (variable === props.meta[0].variable_id) {
-            variableTimes[idxString] = responses[i].data[id].times[timeidx];
-          }
-          if (this.hasValidData('comparand', props) &&
-            variable === props.comparandMeta[0].variable_id
-          ) {
-            comparandTimes[idxString] = responses[i].data[id].times[timeidx];
-          }
-        }
+    // and timestamp to the viewer component DataMap.
+    // The variable and the comparand may have different available timestamps, but will
+    // (aside from variable_id) display the same instance.
+
+    const { start_date, end_date, ensemble_member } = instance;
+    
+    let rasterParamsPromise = getRasterParamsPromise(instance, props.meta);
+    let isolineParamsPromise = getIsolineParamsPromise(instance, props.comparandMeta);
```

Comment:
The following little essay has impact on all the map controllers. I'm putting it here in one place so that I don't distribute and/or repeat the explanations.

Also, I'm not stuck on this being the absolute right way to go, especially in view of the likely need to bind in additional data to the downstream processing functions. But it is worth knowing and considering. I invite your thoughts.

### Principle

It's a good idea to separate things that are inherently promises from those that only coincidentally are. In this case,  `rasterParams` (and `isolinesParams` and `annotatedParams`) are only coincidentally promises, since they are computed from something originally fetched asynchronously. The original fetch is inherently a promise, but the computations downstream of that are only coincidentally promises via the `.then()` chaining. 

A common and helpful practice is to formulate such promise chains as follows:

```
const resultPromise = fetchThing()  // fetchThing returns an inherent promise
  .then(doSomething)
  .then(doSomethingElse)
  .then(doYetAnotherThing);
```

The processing functions themselves do not return promises, they just process values. This separates inherent promises (the result of `fetchThing` and the `then`s) from the downstream processing. 

### Practice

The effect and the point of all this is probably best illustrated with a real example from this code: `getRasterParamsPromise` would be reformulated to

```
rasterParamsPromise = fetchTimeMetadata(datafiles)
   .then(extractTimes)
   .then(makeScalarParams)
   .then(selectRasterPalette)
```

where 
* `fetchTimeMetadata` returns an inherent promise, the `Promise.all()` for the set of metadata fetches in `getTimesPromise`
* `extractTimes` is the anonymous arrow function in `getTimesPromise`
* `makeScalarParams` is the body of `getScalarParamsPromise`
* `selectRasterPalette` is the body of `getRasterParamsPromise`

### Benefits

1. The chaining of `then`s, combined with descriptive function names, makes the processing close to self-documenting, with all the high-level info available in one place at a single glance.
2. The chained functions can be reused in other similar chains. E.g., for the equivalent chains for `getIsolineParamsPromise` and `getAnnotatedParamsPromise`. You've done a nice job factoring these already, but it can get even clearer if you separate promise from non-promise. The code gets slightly more vebose and repetitious (the promise chains are very similar) but it increases clarity.
3. Promises are returned only by functions that inherently must: the `fetch` and the `then`s. The rest are promise-agnostic. This matters because you can create unnecessary tangles (and unrefactorable code) when things that don't _have to be_ concerned with promises _are_. This is an aspect separation of concerns in the context of promises.

### Downsides

1. Verbosity. (Compensated by increased clarity.)
2. The downstream chained functions sometimes need information (values) that they don't get from their chaining inputs. This can be accomplished by partially applying (binding) functions taking the full set of params, yielding a function that takes just the chaining inputs. Not fun, but not really hard either. (You can also pass these extras along the `then` chain, but that usually turns out to be a very bad idea. It would be in this case.)